rfcs/2023-04-18-bundle-check.md

```diff
@@ -0,0 +1,124 @@
+We will achieve the following with this:
+
+3. Ensure our tree-shaking works as expected by adding a check to ensure that unused components are not included in the bundle.
+
+- We will use [size-limit](https://github.com/ai/size-limit) for this
+- size-limit is a tool that helps us to keep an eye on our bundle size. It works by comparing the size of our bundles against a predefined limit. If the size of the bundle exceeds the limit, the build fails.
+- size-limit creates an empty project and imports the component that we want. It then creates a bundle and checks the size of the bundle. If the size of the bundle exceeds the limit, the build fails.
+- This will enable us to limit importing only a small set of components like `{ Button, Text }` and check the bundle size to ensure that tree-shaking is working as intended and the overall bundle size is limited to the imported components.
+- size-limit configuration looks like this:
+```json
+  "size-limit": [
+    {
+      "path": "./build/components/index.web.js",
+      "import": "{ Button, Text }",
+      "limit": "100 kB"
+    }
+  ]
+```
+- We are not using size-limit to do a total bundle size check or master diff because it doesn't support it as well as bundlemon does. Bundlemon has an infra set up that stores master bundle size and diffs it on every PR. We can't do that with size-limit.
+
+# Drawbacks/Constraints
+- Bundlemon is a free service but it also seems unmaintained since Jan 2022.
+- If Bundlemon service goes down, we will not be able to get any master diffs but we will still be able to get the total bundle size limit check.
+- Alternate to using bundlemon is setting up our own infra to store master bundle size and diff it on every PR. This would also be a lot of work and we would have to maintain it.
+- I'd suggest taking the shorter route and using bundlemon for now. 
+- [size-limit](https://github.com/ai/size-limit) is an open source project and is actively maintained.
+- size-limit does not support master diffing. Ref: https://github.com/ai/size-limit/issues/318
+- While setting up size-limit, I encountered certain cases where size-limit was not respecting the config (especially to disable load time checks, disabling gzip and compareWith configs). It works for the use-case we want it to work for. Since it's actively maintained, it could be fixed in the future.
+
+# Alternatives
+- Explored other design systems and how they do it.
+  - Most design systems have their own infra for diffing bundle size with master
+  - Some of the design systems were limited to only checking bundle size limit and not doing any diff
+- Explored internal razorpay projects.
+  - All the projects are only doing bundle size limits and none of them are doing master diffing.
+- ~Discovered [bundlemon](https://github.com/LironEr/bundlemon) while writing this RFC, exploring it as an alternative to bundlewatch~ We will go ahead with bundlemon as it has a few extra features and is actively maintained
```

Comment:
>features and is actively maintained

it isn't actively maintained right?

Reply:
bundle**mon** is actively maintained, bundle**watch** is not actively maintained. We will be using bundle**mon**